python/nav/ipdevpoll/plugins/paloaltoarp.py

```diff
@@ -0,0 +1,150 @@
+#
+
+"""ipdevpoll plugin for fetching arp mappings from Palo Alto firewalls
+
+Add [paloaltoarp] section to ipdevpoll.conf
+add hostname = key to [paloaltoarp] section
+for example:
+[paloaltoarp]
+10.0.0.0 = abcdefghijklmnopqrstuvwxyz1234567890
+
+"""
+
+import xml.etree.ElementTree as ET
+from typing import Dict
+
+from IPy import IP
+from twisted.internet import defer, reactor, ssl
+from twisted.internet.defer import returnValue
+from twisted.web import client
+from twisted.web.client import Agent
+from twisted.web.http_headers import Headers
+
+from nav import buildconf
+from nav.ipdevpoll.plugins.arp import Arp
+
+
+class PaloaltoArp(Arp):
+    configured_devices: Dict[str, str] = {}
+
+    @classmethod
+    def on_plugin_load(cls):
+        """Loads the list of PaloAlto access keys from ipdevpoll.conf into the plugin
+        class instance, so that `can_handle` will be able to answer which devices
+        this plugin can run for.
+        """
+        from nav.ipdevpoll.config import ipdevpoll_conf
+
+        cls._logger.debug("loading paloaltoarp configuration")
+        if 'paloaltoarp' not in ipdevpoll_conf:
+            cls._logger.debug("PaloaltoArp config section NOT found")
+            return
+        cls._logger.debug("PaloaltoArp config section found")
+        cls.configured_devices = dict(ipdevpoll_conf['paloaltoarp'])
+
+    @classmethod
+    def can_handle(cls, netbox):
+        """Return True if this plugin can handle the given netbox."""
+        return (
+            netbox.sysname in cls.configured_devices
+            or str(netbox.ip) in cls.configured_devices
+        )
+
+    @defer.inlineCallbacks
+    def handle(self):
+        """Handle plugin business, return a deferred."""
+
+        api_key = self.configured_devices.get(str(self.netbox.ip), self.netbox.sysname)
+        self._logger.debug("Collecting IP/MAC mappings for Paloalto device")
+
+        mappings = yield self._get_paloalto_arp_mappings(
+            self.netbox.ip,
+            api_key,
+        )
+        if mappings is None:
+            self._logger.info("No mappings found for Paloalto device")
+            returnValue(None)
+
+        yield self._process_data(mappings)
+
+        returnValue(None)
+
+    @defer.inlineCallbacks
+    def _get_paloalto_arp_mappings(self, address: str, key: str):
+        """Get mappings from Paloalto device"""
+
+        arptable = yield self._do_request(address, key)
+        if arptable is None:
+            returnValue(None)
+
+        # process arpdata into an array of mappings
+        mappings = parse_arp(arptable.decode('utf-8'))
+        returnValue(mappings)
+
+    @defer.inlineCallbacks
+    def _do_request(self, address: str, key: str):
+        """Make request to Paloalto device"""
+
+        class SslPolicy(client.BrowserLikePolicyForHTTPS):
+            def creatorForNetloc(self, hostname, port):
+                return ssl.CertificateOptions(verify=False)
```

Comment:
We discussed this at UiT last year. Since we are only extracting ARP data this should be fine as long as the keys and permissions on the Palo Alto device are configured correctly (To only allow for fetching ARP data). But yes we are explicitly turning of TLS certificate verification on all requests made by the plugin.

Reply:
It might be better to implement certificate pinning instead of this solution just in case someone manages to misconfigure their permissions.

Reply:
A MITM attack might also allow for data infiltration, certificate pinning is starting to seem like the best option.

Reply:
To get this PR done with, I would just make note of this fact in the docs.  That way, you could submit a new PR with changes for certificate config options (i.e. switch verification on/off or point to a pinned certificate)